From c5680eb6e6d640f66250c459a807d233a119a3f3 Mon Sep 17 00:00:00 2001
From: Peng Huang <penghuang@chromium.org>
Date: Fri, 24 Nov 2023 21:12:04 +0000
Subject: [PATCH] Fix crash related context lost

The problem is because SkiaOutputDeviceGL::ctor will call MakeCurrent
again for enabling swap timestamp on the GLSurface. It can trigger
context loss. Fix the problem by moving related code to
SkiaOutputSurfaceImplOnGpu::InitializeForGL(). And we will detect the
context lost there.

Bug: 1412675
Change-Id: Ib675190e7b5d2613a46d6ef497d2fc22b06d66a6
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5054411
Reviewed-by: Vasiliy Telezhnikov <vasilyt@chromium.org>
Commit-Queue: Peng Huang <penghuang@chromium.org>
Auto-Submit: Peng Huang <penghuang@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1228882}
---

--- a/components/viz/service/display_embedder/skia_output_device_dcomp.cc
+++ b/components/viz/service/display_embedder/skia_output_device_dcomp.cc
@@ -262,14 +262,6 @@
   capabilities_.supports_delegated_ink = gl_surface_->SupportsDelegatedInk();
   capabilities_.pending_swap_params.max_pending_swaps =
       gl_surface_->GetBufferCount() - 1;
-
-  if (gl_surface_->SupportsSwapTimestamps()) {
-    gl_surface_->SetEnableSwapTimestamps();
-
-    // Changes to swap timestamp queries are only picked up when making current.
-    context_state_->ReleaseCurrent(nullptr);
-    context_state_->MakeCurrent(gl_surface_.get());
-  }
 }
 
 SkiaOutputDeviceDCompGLSurface::~SkiaOutputDeviceDCompGLSurface() {
--- a/components/viz/service/display_embedder/skia_output_device_gl.cc
+++ b/components/viz/service/display_embedder/skia_output_device_gl.cc
@@ -91,14 +91,6 @@
   DCHECK(context_state_);
   DCHECK(gl_surface_);
 
-  if (gl_surface_->SupportsSwapTimestamps()) {
-    gl_surface_->SetEnableSwapTimestamps();
-
-    // Changes to swap timestamp queries are only picked up when making current.
-    context_state_->ReleaseCurrent(nullptr);
-    context_state_->MakeCurrent(gl_surface_.get());
-  }
-
   DCHECK(context_state_->gr_context());
   DCHECK(context_state_->context());
 
--- a/components/viz/service/display_embedder/skia_output_surface_impl_on_gpu.cc
+++ b/components/viz/service/display_embedder/skia_output_surface_impl_on_gpu.cc
@@ -1990,6 +1990,10 @@
       if (!gl_surface_) {
         return false;
       }
+
+      if (gl_surface_->SupportsSwapTimestamps()) {
+        gl_surface_->SetEnableSwapTimestamps();
+      }
     }
 
 #if BUILDFLAG(IS_MAC)
